.github/custom_python_actions/check_cla.py

```diff
@@ -0,0 +1,102 @@
+        self.cla_repo = gh.repository(owner="dfinity", repository="cla")
+        self.cla_link = f"{self.cla_repo.html_url}/blob/main/CLA.md"
+
+    def check_if_cla_signed(self, issue: GHIssue, user: str) -> bool:
+        # check if bot has already left a message to avoid spam
+        for comment in issue.comments():
+            bot_comment = True if comment.user.login == "dfnitowner" else False
+
+        for comment in issue.comments():
+            if comment.user.login == user:
+                agreement_message = messages.USER_AGREEMENT_MESSAGE.format(user)
+                comment_body = comment.body.strip()
+                if comment_body == agreement_message:
+                    print(f"CLA has been agreed to by {user}")
+                    return True
+                else:
+                    print(f"Comment created by {user} does not match CLA agreement.")
+                    if not bot_comment:
+                        issue.create_comment(messages.FAILED_COMMENT)
+        print(f"CLA is pending for {user}")
+        return False
+
+    def get_cla_issue(self, user: str) -> Optional[GHIssue]:
+        for issue in self.cla_repo.issues():
+            if issue.title == f"cla: @{user}":
+                return issue
+        print(f"No CLA issue for {user}")
+        return None  # to make linter happy
+
+    def create_cla_issue(self, user: str) -> GHIssue:
+        user_agreement_message = messages.USER_AGREEMENT_MESSAGE.format(user)
+        issue = self.cla_repo.create_issue(
+            f"cla: @{user}",
+            body=messages.CLA_AGREEMENT_MESSAGE.format(
+                user, self.cla_link, user_agreement_message
+            ),
+            labels=[PENDING_LABEL],
+        )
+        return issue
+
+    def handle_cla_signed(self, issue: GHIssue, user: str) -> None:
+        for label in issue.original_labels:
+            if label.name == APPROVED_LABEL:
+                return
+            elif label.name == PENDING_LABEL:
+                agreement_message = messages.AGREED_MESSAGE.format(user)
+                issue.create_comment(agreement_message)
+                issue.remove_label(PENDING_LABEL)
+                issue.add_labels(APPROVED_LABEL)
+                return
+        print(
+            "No cla labels found - manually check the cla issue to see what state it is in. Exiting program."  # noqa
+        )
+        sys.exit(1)
+
+
+def main() -> None:
+    org = os.environ["GH_ORG"]
+    gh_token = os.environ["GH_TOKEN"]
+    repo = os.environ["REPO"]
+    pr_id = os.environ["PR_ID"]
+
+    gh = github3.login(token=gh_token)
+    pr = gh.pull_request(org, repo, pr_id)
+    user = pr.user.login
+
+    cla = CLAHandler(gh)
+
+    issue = cla.get_cla_issue(user)
+    if not issue:
+        issue = cla.create_cla_issue(user)
+        pr_comment = messages.CLA_MESSAGE.format(user, cla.cla_link, issue.html_url)
+        pr.create_comment(pr_comment)
+
+    cla_signed = cla.check_if_cla_signed(issue, user)
```

Comment:
You could try and check N times if user has signed the agreement with exponential backoff. That would be better user experience IMO.

Reply:
User wouldn't need to re-try the job. Maybe exponential backoff is not a good idea, just periodically checking for 3-5 minutes might be. You create a MR where user needs to agree to CLA and then check every 30 seconds 6-10 times if CLA was signed. If it's signed, then pipeline succeeds and user doesn't need to re-try the failed job.

Reply:
I don't think that's a great solution - I think it's better if it fails vs keeps running for 10+ minutes. Then the user knows that action is needed on their part.

Reply:
I didn't say 10+ minutes. You can see from some PRs what's the time the users usually agreed on CLA and see if you can cover most of the cases in 3-5 minutes. But if you think it's not great solution to see if you can avoid requiring user's action to re-try the failed job, then perhaps just leave it as it is.

Reply:
Yeah, I see your point, but think it follows a pretty typical CI workflow:
1) CI failed
2) make a change
3) trigger CI (either manually or automatically when a code change is pushed)

I think this is a pretty logical workflow, so I don't think it should be too confusing. I'd rather keep things simple in the beginning and then make modifications as pain points are identified.

Reply:
Sorry, didn't check the code yet, just an idea based on this conversation:
we can also have workflow that will run https://github.com/dfinity/cla/issues and automatically trigger re-run of this failed action when the user signed CLA.

Reply:
I was trying to keep the logic in one place and having some logic in this repository, some in the other would introduce more complexity. If you both think this would be bad for the user experience, I can add a spike separate from this PR to see what options are available to re-trigger the workflow once the CLA has been signed.